chapters/VK_EXT_present_timing/PresentTimeInfo.adoc

```diff
@@ -0,0 +1,162 @@
+
+  * pname:targetPresentTime, if non-zero, specifies the earliest time the
+    application wants the image to complete the swapchain's target present
+    stage. A value of zero specifies that the presentation engine may:
+    display the image at any time allowed by the current present mode.
+  * pname:minPresentDuration, if non-zero, specifies the minimum duration in
+    nanoseconds the application wants the image to be displayed to the user.
+    A value of zero specifies that the presentation engine may: display the
+    image for any duration allowed by the current present mode.
+
+The contents of the sname:VkPresentTimeEXT union is interpreted depending on
+the associated sname:VkPresentInfoKHR::pname:pSwapchains member creation
+flags. If ename:VK_SWAPCHAIN_CREATE_ABSOLUTE_TIME_BIT_EXT was used to create
+the pname:swapchain, pname:targetPresentTime is used. If
+ename:VK_SWAPCHAIN_CREATE_RELATIVE_TIME_BIT_EXT was used to create the
+pname:swapchain, pname:minPresentDuration is used.
+
+pname:targetPresentTime is a time in nanoseconds, according to the
+time-domain being used. The presentation engine must: not complete the
+target present stage, specified in the associated swapchain
+slink:VkSwapchainPresentTimingCreateInfoEXT::pname:presentStageTarget, at an
+earlier time.
+
+[NOTE]
+.Note
+====
+Applications that use pname:targetPresentTime to specify an absolute present
+time should: regularly rebase their calculations for their next target time
+on the feedback from flink:vkGetPastPresentationTimingEXT to compensate for
+accumulated precision errors or potential clock drift. It is recommended
+that when targeting the time of a vertical blanking period, applications
+subtract a small delta from pname:targetPresentTime as a margin for error.
+====
+
+include::{generated}/validity/structs/VkPresentTimeEXT.adoc[]
+--
+
+[open,refpage='VkPresentStageTimeEXT',desc='Associate a present stage with a timestamp',type='structs']
+--
+The sname:VkPresentStageTimeEXT structure is defined as:
+
+include::{generated}/api/structs/VkPresentStageTimeEXT.adoc[]
+
+  * pname:stage is a tlink:VkPresentStageFlagsEXT value specifying a present stage.
+  * pname:time is a time in nanoseconds associated with the pname:stage.
+
+include::{generated}/validity/structs/VkPresentStageTimeEXT.adoc[]
+--
+
+[open,refpage='VkPresentStageFlagBitsEXT',desc='Bitmask specifying stages of the image presentation process',type='enums']
+--
+
+Presenting an image to the user typically involves multiple stages. Bits
+which can: be set to specify present stages are:
+
+include::{generated}/api/enums/VkPresentStageFlagBitsEXT.adoc[]
+
+  * ename:VK_PRESENT_STAGE_IMAGE_HANDOFF_BIT_EXT is the stage after which the
+    presentation engine has passed the presentation request of the image to
+    a system-specific entity that is outside the control of the Vulkan
+    implementation. This stage is the only ename:VkPresentStageFlagBitsEXT
+    that is required: to be supported.
```

Comment:
I think the present stages are all confusing what the PE is. The PE is the whole downstream thing. So for example with wayland you have the WSI implementation in the driver, the wayland compositor and then probably a display or maybe even some remote stuff. All of that is the PE if I'm not completely mistaken.

I think `VK_PRESENT_STAGE_IMAGE_HANDOFF_BIT_EXT` here really is supposed to be the time at which `QueuePresent` is called?

`VK_PRESENT_STAGE_IMAGE_LATCHED_BIT_EXT` then means that the PE has latched on to the update to create the presentation for a refresh cycle.

Reply:
> I think the present stages are all confusing what the PE is. The PE is the whole downstream thing.

Agreed, I'll try to reword things better. For example, HANDOFF shouldn't imply that the presentation request has left the PE, merely that the PE moved the request from one "place" to another.

> So for example with wayland you have the WSI implementation in the driver, the wayland compositor and then probably a display or maybe even some remote stuff. All of that is the PE if I'm not completely mistaken.

I was pretty sure that there was at least a distinction between the PE and the hardware that's actually displaying the image, but I'm having doubts now and I can't find the source on that. Did I hallucinate it? @cubanismo might have a better answer.

> I think VK_PRESENT_STAGE_IMAGE_HANDOFF_BIT_EXT here really is supposed to be the time at which QueuePresent is called?

At the very least it would be after the PresentInfo semaphores have been signaled. I don't know enough about the wayland architecture to translate my mental model there, sorry. On Windows, this would be e.g. when the KMD picks up the presentation request.

> VK_PRESENT_STAGE_IMAGE_LATCHED_BIT_EXT then means that the PE has latched on to the update to create the presentation for a refresh cycle.

That does sound better. Thanks!

Reply:
> At the very least it would be after the PresentInfo semaphores have been signaled

Ahh, right, that does make more sense. Shouldn't we call it `VK_PRESENT_STAGE_IMAGE_READY_BIT_EXT` then?

Reply:
> I think the present stages are all confusing what the PE is. The PE is the whole downstream thing.

Yes, I agree.

Returning to my [long comment about three buckets](https://github.com/KhronosGroup/Vulkan-Docs/pull/1364#issuecomment-1505756658):

The action that I had proposed is in bucket one is the action of the the Vulkan present operation becoming "ready" at the PE.  The timestamp for this hand-off can be reliably known to all Vulkan implementations, and with low latency.

The items actions I had proposed are in buckets two and three are actions that occur within the PE, and within the display device. Things that are within the PE are by definition outside of Vulkan. Being outside of Vulkan, there may be challenges to reporting timestamps of these actions. I caution against conflating a useful present timing extension with insistence of reports for bucket two and bucket three timestamps that cover one user's configuration but not other user's configurations.  Such an attempt does (and has?) caused an amount of rathole discussion causing delay.  I'd like for this first extension to include bucket one, and for a separate effort to contemplate the controversial bucket two items.

The timestamps of bucket two actions can be known to Vulkan. There might be a performance penalty that is incurred when Vulkan is required to measure these timestamps. These timestamps might be "variable" such that timeline discontinuity might confuse an application into buggy behavior. I suggested making reports for these actions be opt-into by the application. In the aforementioned long comment, I also described that we might need a "topology" API surface in order to meaningfully describe all types of system configurations that are already in use today. (Just because one user doesn't care about clone does not mean it's OK to ship an extension that is fundamentally broken for configs that have cloned displays.)

The timestamps of bucket three cannot be known to Vulkan in a meaningful way. For some system configurations, "first photons out" is in this bucket.

Reply:
I've reworded a bit the present stages definitions. I'm still not entirely satisfied with that part though. Maybe I should add a .NOTE section which paraphrases some of the comments from @jjulianoatnv, particularly about the topology of the PE.

> Ahh, right, that does make more sense. Shouldn't we call it VK_PRESENT_STAGE_IMAGE_READY_BIT_EXT then?

@swick One of my goals with HANDOFF was to have a somewhat generic stage that could be used for the more simple class of non-trivial PE topologies, e.g. "nested" PEs. The current spec allows for the same stage enum to be reported multiple times in the presentation feedback data. An example that came up (raised by @versalinyaa, I believe) was an application using Vulkan to present images inside a virtual machine and does not want the timing of the handoff from the VM to the host, but from the host to whatever system is responsible for the final display. Multiple HANDOFF timings could then be reported for that scenario (security implications notwithstanding). This goes against the idea of the "first bucket" @jjulianoatnv expresses though; or rather, the first HANDOFF of the feedback data is implicitly the "first bucket", and the next ones are all "second bucket". I'm open to changing HANDOFF to a more strict operation and leaving what I described to a followup extension, or adding a separate thing entirely.